syntax = "proto3";

package protobuf;

message Transaction {
    bytes     execer    = 1;
    bytes     payload   = 2;
    Signature signature = 3;
    int64     fee       = 4;
    int64     expire    = 5;
    //随机ID，可以防止payload 相同的时候，交易重复
    int64 nonce = 6;
    //对方地址，如果没有对方地址，可以为空
    string to         = 7;
    int32  groupCount = 8;
    bytes  header     = 9;
    bytes  next       = 10;
}

message Signature {
    int32 ty     = 1;
    bytes pubkey = 2;
    //当ty为5时，格式应该用RingSignature去解析
    bytes signature = 3;
}

// message for execs.coins
message CoinsAction {
    oneof value {
        AssetsTransfer       transfer       = 1;
        AssetsWithdraw       withdraw       = 4;
        AssetsGenesis        genesis        = 2;
        AssetsTransferToExec transferToExec = 5;
    }
    int32 ty = 3;
}
// assert transfer struct
message AssetsGenesis {
    int64  amount        = 2;
    string returnAddress = 3;
}

message AssetsTransferToExec {
    string cointoken = 1;
    int64  amount    = 2;
    bytes  note      = 3;
    string execName  = 4;
    string to        = 5;
}

message AssetsWithdraw {
    string cointoken = 1;
    int64  amount    = 2;
    bytes  note      = 3;
    string execName  = 4;
    string to        = 5;
}

message AssetsTransfer {
    string cointoken = 1;
    int64  amount    = 2;
    bytes  note      = 3;
    string to        = 4;
}

message Asset {
    string exec   = 1;
    string symbol = 2;
    int64  amount = 3;
}

message Transactions {
    repeated Transaction txs = 1;
}
